server/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -683,14 +683,8 @@ private Snapshot newMultiSnapshot(TranslogSnapshot[] snapshots) throws IOExcepti
     }
 
     private Stream<? extends BaseTranslogReader> readersAboveMinSeqNo(long minSeqNo) {
-        assert readLock.isHeldByCurrentThread() || writeLock.isHeldByCurrentThread() :
-            "callers of readersAboveMinSeqNo must hold a lock: readLock ["
-                + readLock.isHeldByCurrentThread() + "], writeLock [" + readLock.isHeldByCurrentThread() + "]";
-        return Stream.concat(readers.stream(), Stream.of(current))
-            .filter(reader -> {
-                final long maxSeqNo = reader.getCheckpoint().maxSeqNo;
-                return maxSeqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || maxSeqNo >= minSeqNo;
-            });
+        assert readLock.isHeldByCurrentThread();
+        return Stream.concat(readers.stream(), Stream.of(current)).filter(reader -> reader.getCheckpoint().maxSeqNo >= minSeqNo);
```

Comment:
This method is consumed by peer-recovery and resync. Peer-recovery already excluded operations without sequence numbers (in #27580); therefore we can safely exclude readers without sequence numbers.